### ISSUE TITLE
[BUG] PyYAML doesn't list Python 3.8 as a compatible runtime

### ISSUE THREAD

**Issue by dacut** [keithrozario/Klayers#107]
**Describe the bug**
The current version of the PyYAML layer doesn't report Python 3.8 as a compatible version, despite being built for it. (This was true of the regions I checked: us-*, ca-*)

**Screenshots**
```
aws lambda get-layer-version-by-arn --arn arn:aws:lambda:us-west-2:770693421928:layer:Klayers-python38-PyYAML:3
{
    "Content": {
        "Location": "https://awslambda-...df",
        "CodeSha256": "gYVoWVlhcAYOej80J1Cxo23o20fRaBufPBBZ1vN5tHI=",
        "CodeSize": 103715
    },
    "LayerArn": "arn:aws:lambda:us-west-2:770693421928:layer:Klayers-python38-PyYAML",
    "LayerVersionArn": "arn:aws:lambda:us-west-2:770693421928:layer:Klayers-python38-PyYAML:3",
    "Description": "PyYAML==5.3.1 | 55dae5eec93bd08d60198103ea73ca1f32d48d91269e80b2487fe279b838f329",
    "CreatedDate": "2020-03-23T02:00:38.261+0000",
    "Version": 3,
    "CompatibleRuntimes": [
        "python3.6",
        "python3.7"
    ],
    "LicenseInfo": "MIT"
}
```

Note that CompatibleRuntimes lists Python 3.6 and 3.7, but not Python 3.8.

**Layer Version ARN:**
Probably all of the PyYAML ARNs, but these were the ones I checked:
```
arn:aws:lambda:ca-central-1:770693421928:layer:Klayers-python38-PyYAML:3
arn:aws:lambda:us-east-1:770693421928:layer:Klayers-python38-PyYAML:3
arn:aws:lambda:us-east-2:770693421928:layer:Klayers-python38-PyYAML:3
arn:aws:lambda:us-west-1:770693421928:layer:Klayers-python38-PyYAML:3
arn:aws:lambda:us-west-2:770693421928:layer:Klayers-python38-PyYAML:3
```

**Framework:**
Terraform and raw Python.

**Additional context**
Downloading the ZIP file and re-uploading it to my account, tagging it with Python 3.8, and running it seems to work fine.

**Comment by keithrozario** [comment]
Thanks, will take a look. 

It appears that the compatible runtime list is actually buggy, and should be updated. 

Will correct this.

Do you know how one might determine if a package is compatible with a runtime?

**Comment by dacut** [comment]
tl;dr: Best bet is to `pip3 install` it on each platform and either

- Create separate versions for each platform (PyYAML-py36/-py37/-py38...), or
- Merge the resulting builds into a single ZIP, testing if each file has identical contents when you end up with the same paths.

Long story: There's not a good way to do this.

Pure-Python packages _should_ be compatible across minor versions -- Python guarantees this will be possible, so a package built for 3.4 will run on 3.8. However, there's no guarantee that a package actually is without testing it. (I only support Python 3.6+ on my packages, and sometimes I get it wrong since I usually build on 3.8.)

Binary packages only work for a given minor version + OS.

PyYAML doesn't fall cleanly into either category. It normally attempts to build a binary helper, but will fall back to a pure-Python version (which appears to be the case with how yours was built) if the dependencies it needs to build the binary aren't installed. For example, I have a `_yaml.cpython-38-darwin.so` helper library in my virtualenv on my laptop; there's no equivalent `_yaml.cpython-38-linux.so` in your package, however.

(And even Python doesn't quit get this quite right, because `linux` is too generic -- an .so built for Ubuntu may not run on CentOS due to dependency differences, and almost certainly will have issues on Alpine due to libc differences.)

**Comment by keithrozario** [comment]
Thanks. I do some digging to see if we can actually figure out the compatible runtimes.

If truly difficult, then perhaps I'll just hard-code everything to Python3.8 :(

**Comment by dacut** [comment]
How are you getting the list of compatible runtimes currently?

I'm surprised that Python 3.8 isn't included in each of them given the name of the layer (Kayers-python38-...) -- sounds like you're building them on Python 3.8 given the name.

Aside: looks like you're in Singapore -- hope things are well over there! (I'm near Seattle, US; 15 hours behind... my day is just getting started...)

**Comment by keithrozario** [comment]
Thanks for asking man. Singapore is alright for now, we might even open the borders to Malaysia (where I'm from).

Hope everything is good over there in Seattle.

The list of compatible runtimes is just defaulted to a fix value for all packages. It doesn't really mean anything for now -- it just hard codes it to the values.

**Comment by keithrozario** [comment]
It looks like Lambda Layers can't have their compatible runtimes field updated. The only way is to publish a new layer version altogether.

There seems to be no way to update this, does it cause you any issues? I can manually trigger a version update for PyYAML to fix this if you wish.

**Comment by dacut** [comment]
I’ve already copied it and made modifications to my version directly. No issues for me. :-) I just filed this do you would be aware. Thanks!

On Aug 13, 2020, at 4:07 AM, Keith Rozario <notifications@github.com> wrote:

﻿

It looks like Lambda Layers can't have their compatible runtimes field updated. The only way is to publish a new layer version altogether.

There seems to be no way to update this, does it cause you any issues? I can manually trigger a version update for PyYAML to fix this if you wish.

**Comment by keithrozario** [comment]
Awesome man. Closing this.
